Sources/NIO/Selector.swift

```diff
@@ -41,6 +41,192 @@ private extension Optional {
     }
 }
 
+/// Represents IO events NIO might be interested in. `SelectorEventSet` is used for two purposed: To express interest
+/// in a given event set and for notifications about an IO event set that has occured.
+///
+/// For example, if you were interested in reading and writing data from/to a socket and also obviously if the socket
+/// receives a connection reset, express interest with `[.read, .write, .reset]`.
+/// If then suddenly the socket becomes both readable and writable, the eventing mechanism will tell you about that
+/// fact using `[.read, .write]`.
+struct SelectorEventSet: OptionSet, Equatable {
+
+    typealias RawValue = UInt8
+
+    let rawValue: UInt8
+
+    /// It's impossible to actually register for no events, therefore `_none` should only be used to bootstrap a set
+    /// of flags or to compare against spurious wakeups.
+    static let _none = SelectorEventSet(rawValue: 0)
+
+    /// Connection reset or other errors.
+    static let reset = SelectorEventSet(rawValue: 1 << 0)
+
+    /// EOF at the read/input end of a `Selectable`.
+    static let readEOF = SelectorEventSet(rawValue: 1 << 1)
+
+    /// Interest in/availability of data to be read
+    static let read = SelectorEventSet(rawValue: 1 << 2)
+
+    /// Interest in/availability of data to be written
+    static let write = SelectorEventSet(rawValue: 1 << 3)
+
+    init(rawValue: SelectorEventSet.RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+/// Represents the `kqueue` filters we might use:
+///
+///  - `except` corresponds to `EVFILT_EXCEPT`
+///  - `read` corresponds to `EVFILT_READ`
+///  - `write` corresponds to `EVFILT_WRITE`
+private struct KQueueEventFilterSet: OptionSet, Equatable {
+    typealias RawValue = UInt8
+
+    let rawValue: UInt8
+
+    static let _none = KQueueEventFilterSet(rawValue: 0)
+    static let except = KQueueEventFilterSet(rawValue: 1 << 1)
+    static let read = KQueueEventFilterSet(rawValue: 1 << 2)
+    static let write = KQueueEventFilterSet(rawValue: 1 << 3)
+
+    init(rawValue: RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+/// Represents the `epoll` filters/events we might use:
+///
+///  - `hangup` corresponds to `EPOLLHUP`
+///  - `readHangup` corresponds to `EPOLLRDHUP`
+///  - `input` corresponds to `EPOLLIN`
+///  - `output` corresponds to `EPOLLOUT`
+///  - `error` corresponds to `EPOLLERR`
+private struct EpollFilterSet: OptionSet, Equatable {
+    typealias RawValue = UInt8
```

Comment:
Do you want to move the typealias out of the struct as you define the same multiple times.

Reply:
no I can't, `OptionSet` requires it. I could remove that line entirely and rely on the inference but there'll be changes in that area so let's be explicit.